### PR TITLE
usb: dfu: disable USB DFU uploading by default

### PR DESCRIPTION
**Pull request by jfischer-no**
usb: do not add ZLP if lenght to be transmitted is just zero

It is possible that the actual data stage length of
the control transfer is zero, in that case we do not
need an additional ZLP packet.

This fixes a problem with USB DFU, where after an upload
the device is no longer responsive if upload size is
multiple of control endpoint MPS.

---

usb: dfu: disable USB DFU uploading by default

Firmware uploading to the host may not always be desired,
disable it by default.

Enable it for existing USB DFU sample to keep
the usual behavior, add note about new option to README.rst.

### CODE REVIEW

File: subsys/usb/class/dfu/Kconfig [hunk 1]
```diff
@@ -29,4 +29,10 @@ config USB_DFU_DEFAULT_POLLTIMEOUT
 	help
 	  Default value for bwPollTimeout (in ms)
 
+config USB_DFU_ENABLE_UPLOAD
+	bool "Enable firmware uploading to the host"
+	help
+	  Enabling this option can lead to uploading of decrypted firmware
```
Comment (ddavidebor) [inline subsys/usb/class/dfu/Kconfig:35]: Hi,  this description is not very clear

Comment (nvlsianpu) [reply subsys/usb/class/dfu/Kconfig:35]: Enabling this option allows to download the image by the host. Be aware that the primary image (executable one) is always plain-text despite the image encryption is enabled.

**Review by nvlsianpu — COMMENTED** [review]
This patch looks as I expected.

File: subsys/usb/class/dfu/Kconfig (hunk 1 above)
Comment (de-nordic) [reply subsys/usb/class/dfu/Kconfig:35]: Maybe do the `Security Risk: Enabling this option ...` to make it more clear

Comment (jfischer-no) [reply subsys/usb/class/dfu/Kconfig:35]: Is is upload / uploading, device uploads the firmware to the host, downloading is _from host to device_.  Clearly described according to the USB DFU spec.

Comment (nvlsianpu) [reply subsys/usb/class/dfu/Kconfig:35]: @jfischer-no Yes, let use `upload` word:
Security Risk: Enabling this option allows to upload the image to the host. Be aware that the primary image (executable one) is always plain-text despite the image encryption is enabled.

Comment (jfischer-no) [reply subsys/usb/class/dfu/Kconfig:35]: What about single slot?

Comment (de-nordic) [reply subsys/usb/class/dfu/Kconfig:35]: Single slot update works with mcuboot serial recovery and allows decryption in place on serial upload. This means that allowing download is risk in application running from single slot if it has been uploaded as encrypted.

Comment (de-nordic) [reply subsys/usb/class/dfu/Kconfig:35]: Unfortunately we do not, yet,  have a method to get information from on mcuboot configuration to resolve this at application runtime. Application also does not know whether it has been uploaded encrypted.

Comment (nvlsianpu) [reply subsys/usb/class/dfu/Kconfig:35]: > Decrypted firmware is correct term in this case, not plain-text.

OK

**Review by hackwerken — COMMENTED** [review]
Good thing this is being fixed :)

Note that the device can let the host know it is not capable of uploading. See https://github.com/zephyrproject-rtos/zephyr/blob/af2f936d8d0000e143faa32e5cf997160082554c/subsys/usb/class/dfu/usb_dfu.c#L104

And USB DFU Specification page 13

File: subsys/usb/class/dfu/usb_dfu.c [hunk 2]
```diff
@@ -427,6 +427,13 @@ static int dfu_class_handle_to_host(struct usb_setup_packet *setup,
 		LOG_DBG("DFU_UPLOAD block %d, len %d, state %d",
 			setup->wValue, setup->wLength, dfu_data.state);
 
+		if (!IS_ENABLED(CONFIG_USB_DFU_ENABLE_UPLOAD)) {
```
[the remark this answers is not in the document]
Comment (jfischer-no) [reply subsys/usb/class/dfu/usb_dfu.c:430]: No, my wish would be that underlying flash API simply has no access to this area if not enabled.

File: subsys/usb/class/dfu/Kconfig (hunk 1 above)
Comment (jfischer-no) [reply subsys/usb/class/dfu/Kconfig]: Updated description

**Review by de-nordic — APPROVED** [review]
Looks OK.
Tested on nrf52840dk_nrf52840 with the upload enabled and disabled and works OK.

**Review by carlescufi — APPROVED** [review]

**Comment by carlescufi** [comment]
@nvlsianpu can you please take a look?

**Review by nvlsianpu — APPROVED** [review]

**Review by ceolin — APPROVED** [review]

**Pull request merged**